### PR TITLE
docstore: merge driver specific tests into conformance tests

### PR DESCRIPTION
**Pull request by shantuo**
Updates #2422

This is part I of the change, which remove some of the action tests, if the idea is accepted, I'll continue with part II which will contain query tests.

### docstore
* TestToDriverMods: merged into `TestActionsDo` as the correctionness of the mods should be covered in conformance tests.

### gcpfirestore

* TestNewGetRequest: no need to test the implementation of how batch get requests are constructed, conformance tests should cover various cases.

### memdocstore

* TestUpdateEncodesValues: no need to test `update` encodes values, this should be tested in the codec tests.
* TestMissingKeyCreateFailsWithKeyFunc: this is testing docstore.ActionList.Do and should be replaced by portable API and conformance tests (Done in #2589).
* TestExampleInDoc: should be covered by conformance tests.

### docstore

* Added struct test cases.
* TestToDriverActionsErrors: changed to `TestActionsDo` to use the exported method `Do`.

### conformance tests

* Merged `MultipleActions` and `UnorderedActions` since we only have unordered mode now.
* Added `testExampleInDoc`.

### CODE REVIEW

**Comment by jba** [comment]
> * TestToDriverMods: merged into `TestActionsDo` as the correctionness of the mods should be covered in conformance tests.

I'm OK with this change, but it has nothing to do with the conformance tests. 

> * TestResourceIDRegexp: no need to test the resource ID regex.

Why not? Regexes are hard to get right. It's important to test the edge cases.

> * TestUpdateEncodesValues: no need to test `update` encodes values, this should be tested in the codec tests.

But the codec tests check that various things are encoded correctly, not that encoding is performed in the first place.

Please do the following: 
1. Remove the logic from memdocstore that encodes updates.
2. Run all the tests and verify that some test breaks, *and* that it breaks in a way that clearly points to this bug.

The second part is important: if a test breaks but it's hard to see why, then it can be useful to have a redundant test that targets the exact problem.


> * TestMissingKeyCreateFailsWithKeyFunc: this is testing docstore.ActionList.Do and should be replaced by portable API and conformance tests

Please do the same exercise as above.

In general, don't remove a test unless you can prove that it's covered elsewhere.

> * TestExampleInDoc: should be covered by conformance tests.

Yes, it's probably a good idea to make sure this example works for all providers. But don't remove the test until then.

> Changed portable API and conformance tests

You haven't changed any conformance tests. The file `drivertest.go` is unmodified.

**Comment by shantuo** [comment]
> > * TestToDriverMods: merged into `TestActionsDo` as the correctionness of the mods should be covered in conformance tests.
> 
> I'm OK with this change, but it has nothing to do with the conformance tests.
> 
The conformance part is talking about testing the results of updates.
> > * TestResourceIDRegexp: no need to test the resource ID regex.
> 
> Why not? Regexes are hard to get right. It's important to test the edge cases.
> 
I was thinking that would fail anyway if the user passes in an invalid ID. But I guess it makes sense to keep the test since we don't want to reject good ones?
> > * TestUpdateEncodesValues: no need to test `update` encodes values, this should be tested in the codec tests.
> 
> But the codec tests check that various things are encoded correctly, not that encoding is performed in the first place.
> 
> Please do the following:
> 
> 1. Remove the logic from memdocstore that encodes updates.
> 2. Run all the tests and verify that some test breaks, _and_ that it breaks in a way that clearly points to this bug.
> 
> The second part is important: if a test breaks but it's hard to see why, then it can be useful to have a redundant test that targets the exact problem.
> 
Yes I noticed that, the tests will break if the mods are not encoded, it either returns invalidArgument (for bad increment value), or has type mismatch (for directly updating numbers). I've also added test cases to testUpdate for that, but in a different PR #2594.
> > * TestMissingKeyCreateFailsWithKeyFunc: this is testing docstore.ActionList.Do and should be replaced by portable API and conformance tests
> 
> Please do the same exercise as above.
> 
Yes I do have tests for that but blocked by #2589, can I make a separate PR to address that together? The test for this case is not working as I expected now.
> In general, don't remove a test unless you can prove that it's covered elsewhere.
> 
> > * TestExampleInDoc: should be covered by conformance tests.
> 
> Yes, it's probably a good idea to make sure this example works for all providers. But don't remove the test until then.
> 
Yes, I added this example type to the portable API test, and I plan to also test it in the conformance tests.
> > Changed portable API and conformance tests
> 
> You haven't changed any conformance tests. The file `drivertest.go` is unmodified.
I wasn't include that file since some tests are failing due to #2589 , can I address that in a separate PR?

**Comment by shantuo** [comment]
PTAL

**Comment by jba** [comment]
It sounds like #2589 should be addressed before continuing with this PR.

**Comment by shantuo** [comment]
PTAL

File: docstore/drivertest/drivertest.go [hunk 1]
```diff
@@ -1819,6 +1764,76 @@ func testActionsOnStructWithoutRevision(t *testing.T, coll *ds.Collection) {
...
+			t.Fatal(err)
+		}
+	}
+	checkFieldEqual := func(got, want interface{}, field string) {
+		t.Helper()
+		dg, err := driver.NewDocument(got)
+		must(err)
+		dw, err := driver.NewDocument(want)
+		must(err)
+		fvg, err := dg.GetField(field)
```
Comment (jba) [inline docstore/drivertest/drivertest.go]:
```
fvg, err := MustDocument(got).GetField(field)
```
etc.
